[libc] Use `${libc_opt_high_flag}` instead of `-O3`

petrhosek opened the pull request:
This is preferable since `${libc_opt_high_flag}` will be set correctly for the compiler used.

petrhosek committed:
[libc] Use `${libc_opt_high_flag}` instead of `-O3`

This is preferable since `${libc_opt_high_flag}` will be set correctly
for the compiler used.

jhuber6 commented:
Do we want to override debug builds?

petrhosek commented:
> Do we want to override debug builds?

We talked about it in he libc meeting today and agreed that we shouldn't be overriding the flag and instead rely on whatever user sets in CMake (except for cases where `-O3` is absolutely necessary). I plan on doing that in a follow up PR, this change is just an initial cleanup.

lntue commented:
Can you take a look at the Windows presubmit failure?

nickdesaulniers commented on libc/src/math/generic/CMakeLists.txt:
```diff
@@ -534,7 +534,7 @@ add_entrypoint_object(
     libc.src.__support.macros.properties.types
   COMPILE_OPTIONS
-    -O3
+    ${libc_opt_high_flag}
```
Is there a way to avoid having to spell this out again and again and again for each entry point? I don't see why the math.h functions should be built with a different optimization level than the rest of the libc.  The codebase should use the same `-O*` flag consistently everywhere.

lntue replied:
I'm ok with removing the separate optimization level for math functions, but we do need to add build bots with different optimization levels to make sure that math functions are still correct.

nickdesaulniers replied:
Why would they be incorrect based on optimization level?

lntue replied:
-O3 was notorious for aggressive optimizations, especially with gcc.  Nowadays -O2 and -O3 are probably close, but there are still differences, and it's probably safer to make sure different optimizations do not change the floating point computations' outcomes.

nickdesaulniers replied:
We should key off what optimization levels are set for `CMAKE_BUILD_TYPE` and not override the users intent. If users hack up their configs to use different optimization levels, let them report bugs upstream to us.

petrhosek commented:
> Can you take a look at the Windows presubmit failure?

Looks like the optimization flag is being passed to both the compiler and the linker. That's fine on Linux or macOS since `clang` is invoked as the linker and it'll consume the optimization flag, but on Windows we invoke the linker directly and `lld-link` (or `link`) doesn't handle `/O2`. I don't know why this works now, it's possible that `lld-link` just consumes and ignores `-O3`.

petrhosek committed:
Avoid applying compile options to the linker

Compile options aren't the same as link options and on some platforms
like Windows where the linker is invoked directly they're distinctly
different.

nickdesaulniers commented:
One thing to triple check is that when cross compiling (`-DLIBC_TARGET_TRIPLE=aarch64-linux-gnu`) and using lld (`-DLLVM_ENABLE_LLD=ON`), do we still pass `--target=aarch64-linux-gnu` to the linker?

petrhosek committed:
Pass common compile options to the linker

petrhosek committed:
Actually set link options

petrhosek commented:
Looks like all the tests are passing now.
